core/shared/src/main/scala/scalaz/zio/RefM.scala

```diff
@@ -0,0 +1,87 @@
+// Copyright (C) 2017-2018 John A. De Goes. All rights reserved.
+package scalaz.zio
+
+/**
+ * A mutable atomic reference for the `IO` monad. This is the `IO` equivalent of
+ * a volatile `var`, augmented with atomic effectful operations, which make it
+ * useful as a reasonably efficient (if low-level) concurrency primitive.
+ *
+ * Unlike `Ref`, `RefM` allows effects in atomic operations, which makes the
+ * structure slower but more powerful than `Ref`.
+ *
+ * {{{
+ * for {
+ *   ref <- RefM(2)
+ *   v   <- ref.update(_ + putStrLn("Hello World!").attempt.void *> IO.now(3))
+ *   _   <- putStrLn("Value = " + v) // Value = 5
+ * } yield ()
+ * }}}
+ */
+final class RefM[A] private (private val value: Ref[A], private val queue: Queue[RefM.Bundle[A, _]])
```

Comment:
I suggest if we define value and queue without `val` they would be inaccessible :
 ```scala
final class RefM[A] private (value: Ref[A], queue: Queue[RefM.Bundle[A, _]])
```